faiss/IndexHNSW.cpp

```diff
@@ -647,6 +647,155 @@ IndexHNSWFlat::IndexHNSWFlat(int d, int M, MetricType metric)
     is_trained = true;
 }
 
+IndexHNSWFlat::IndexHNSWFlat(int d, int M, IndexFlat* storage)
+        : IndexHNSW(storage, M) {
+    own_fields = true;
+    is_trained = true;
+}
+
+/**************************************************************
+ * IndexHNSWFlatPanorama implementation
+ **************************************************************/
+
+namespace {
+void compute_cum_sums(
+        const float* x,
+        float* dst_cum_sums,
+        int d,
+        int n_levels,
+        int level_width) {
+    // Iterate backwards through levels, accumulating sum as we go.
+    // This avoids computing the suffix sum for each vector, which takes
+    // extra memory.
+
+    float sum = 0.0f;
+    dst_cum_sums[n_levels] = 0.0f;
+    for (int level = n_levels - 1; level >= 0; level--) {
+        int start_idx = level * level_width;
+        int end_idx = std::min(start_idx + level_width, d);
+        for (int j = start_idx; j < end_idx; j++) {
+            sum += x[j] * x[j];
+        }
+        dst_cum_sums[level] = sqrt(sum);
```

Comment:
is there a reason to do a sqrt (ie. use L2 distance instead of squared L2 distance as usual)?